### PR TITLE
[Feature] Add HamiltonianEvolution, Support py3.12

### PR DESCRIPTION
**Pull request by dominikandreasseitz**
Fixes https://github.com/pasqal-io/horqrux/issues/8

### CODE REVIEW

File: tests/test_analog.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+
+def test_hamevo_single() -> None:
+    n_qubits = 4
+    t_evo = jnp.pi / 4
+    hamevo = HamiltonianEvolution(tuple([i for i in range(n_qubits)]))
+    psi = uniform_state(n_qubits)
+    psi_star = apply_gate(psi, hamevo, {"hamiltonian": Hamiltonian, "time_evolution": t_evo})
+    result = overlap(psi_star, psi)
+    assert jnp.isclose(result, 0.5)
```
Comment (jpmoutinho) [inline tests/test_analog.py:22]: Can you also replicate the random one from PyQ?

```
def Hamiltonian_general(n_qubits: int = 2, batch_size: int = 1) -> torch.Tensor:
    H_batch = torch.zeros((2**n_qubits, 2**n_qubits, batch_size), dtype=torch.cdouble)
    for i in range(batch_size):
        H_0 = torch.randn((2**n_qubits, 2**n_qubits), dtype=torch.cdouble)
        H = (H_0 + torch.conj(H_0.transpose(0, 1))).cdouble()
        H_batch[..., i] = H
    return H_batch

@pytest.mark.parametrize("n_qubits, batch_size", [(2, 1), (4, 2)])
def test_hamevo_general(n_qubits: int, batch_size: int) -> None:
    H = Hamiltonian_general(n_qubits, batch_size)
    t_evo = torch.rand(1, dtype=torch.float64)
    hamevo = pyq.HamiltonianEvolution(tuple([i for i in range(n_qubits)]), n_qubits)
    psi = pyq.random_state(n_qubits, batch_size)
    psi_star = hamevo(H, t_evo, psi)
    assert is_normalized(psi_star)
```

Comment (jpmoutinho) [reply tests/test_analog.py:22]: Note you can simply do
```
def is_normalized(state):
      return equivalent_state(state, state)
```
if you wish to add it to the `utils.py`

**Review by jpmoutinho — COMMENTED** [review]
Thanks @dominikandreasseitz, looks good. I would also open an issue about adding a version just using `exp` and not `expm` for diagonal Hamiltonians.

**Review by jpmoutinho — APPROVED** [review]

File: tests/test_analog.py (hunk 1 above)
Comment (dominikandreasseitz) [reply tests/test_analog.py]: added

**Pull request merged**